Support ConvTranspose2d BatchNorm2d fusion

triple-mu opened the pull request:
Testing pass:

``` python
import torch
import torch.nn as nn
from mmcv.cnn.utils.fuse_conv_bn import _fuse_conv_bn
import random

random.seed(0)


@torch.no_grad()
def test_fusion():
    in_channel = random.randint(1, 100)
    out_channels = random.randint(1, 100)
    kernel_size = random.randint(1, 7)
    stride = random.randint(1, 2)
    padding = random.randint(0, 1)
    height = random.randint(32, 640)
    width = random.randint(32, 640)
    batch = random.randint(1, 32)
    conv = nn.Conv2d(in_channel, out_channels, kernel_size=kernel_size, stride=stride, padding=padding)
    cbn = nn.BatchNorm2d(out_channels)
    conv.eval()
    cbn.eval()

    deconv = nn.ConvTranspose2d(in_channel, out_channels, kernel_size=kernel_size, stride=stride, padding=padding)
    dbn = nn.BatchNorm2d(out_channels)
    deconv.eval()
    dbn.eval()

    x = torch.randn(batch, in_channel, height, width)

    conv_out = cbn(conv(x))
    fconv = _fuse_conv_bn(conv, cbn)
    fconv_out = fconv(x)
    torch.testing.assert_close(conv_out, fconv_out, rtol=1e-5, atol=1e-5)

    dconv_out = dbn(deconv(x))
    fdeconv = _fuse_conv_bn(deconv, dbn)
    fdconv_out = fdeconv(x)
    torch.testing.assert_close(dconv_out, fdconv_out, rtol=1e-5, atol=1e-5)


def main():
    for _ in range(100):
        test_fusion()


if __name__ == '__main__':
    main()

```

triple-mu committed:
Support ConvTranspose2d BatchNorm2d fusion

HAOCHENYE reviewed (commented):
Hi, the corresponding unit test should also be updated to make sure the implementation is right.

zhouzaida committed:
[Refactor] Refactor the pytorch and cuda combination for testing in CI (#2693)

* Refactor CI

* Upload coverage to Codecov

* update

* update

* update

* update

* test

* test

* fix typo

* fix

* fix

* fix

* fix circleci

* fix circleci

* revert comments

* add verpose

* install cuda

install cuda

install cuda

install cuda

install cuda

install cuda

install cuda

fix env

fix env

fix env

fix env

fix

fix

fix

fix

fix

fix

fix

fix

fix

fix

fix

fix

fix

fix circleci

fix circleci

remove cuda11.1 test from windows

update ci

update ci

update ci

update ci

update ci

update ci

update ci

update ci

* do not test python3.11

* revert comments

CokeDong committed:
[Fix] Fix parrots CI (#2737)

* Chery pick from PR #2515 and #2684

* Fix

zhouzaida committed:
Remove two outdated files (#2740)

zhouzaida committed:
Update the supported torch and cuda versions in README (#2687)

* Update the supported torch and cuda versions in README

* update version.json

* use ubuntu-22.04

* refine

zhouzaida committed:
[Docs] Describe branch changes (#2741)

* [Docs] Describe branch changes

* fix typo

* Update English

* refine

zhouzaida committed:
Remove .owners.yml (#2743)

jiangjiang-coder66 committed:
Pick MLU modifications from master (1.x) to main (2.x) (#2704)

* [Feature] Support Voxelization with cambricon MLU device (#2500)

* [Feature] Support hard_voxelize with cambricon MLU backend

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Feature](bangc-ops): add voxelization op

* [Enhance] Optimize the performace of ms_deform_attn for MLU device (#2510)

* ms_opt

* ms_opt

* ms_opt

* ms_opt

* ms_opt

* [Feature] ms_deform_attn performance optimization

* [Feature] ms_deform_attn performance optimization

* [Feature] ms_deform_attn performance optimization

* [Feature] Support ball_query with cambricon MLU backend and mlu-ops library. (#2520)

* [Feature] Support ball_query with cambricon MLU backend and mlu-ops library.

* [Fix] update operator data layout setting.

* [Fix] add cxx compile option to avoid symbol conflict.

* [Fix] fix lint errors.

* [Fix] update ops.md with info of ball_query support by MLU backend.

* [Feature] Fix typo.

* [Fix] Remove print.

* [Fix] get mlu-ops from MMCV_MLU_OPS_PATH env.

* [Fix] update MMCV_MLU_OPS_PATH check logic.

* [Fix] update error info when failed to download mlu-ops.

* [Fix] check mlu-ops version matching info in mmcv.

* [Fix] revise wrong filename.

* [Fix] remove f.close and re.

* [Docs] Steps to compile mmcv-full on MLU machine (#2571)

* [Docs] Steps to compile mmcv-full on MLU machine

* [Docs] Adjust paragraph order

* Update docs/zh_cn/get_started/build.md

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

* Update docs/zh_cn/get_started/build.md

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

* Update docs/en/get_started/build.md

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

* Update docs/en/get_started/build.md

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

* [Docs] Modify the format

---------

Co-authored-by: budefei <budefei@cambricon.com>
Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

* [Fix] Fix tensor descriptor setting in MLU ball_query. (#2579)

* [Feature] Add MLU support for Sparse Convolution op (#2589)

* [Feature] Add sparse convolution MLU API

* [Feature] update cpp code style

* end-of-file

* delete libext.a

* code style

* update ops.md

---------

Co-authored-by: budefei <budefei@cambricon.com>

* [Enhancement] Replace the implementation of deform_roi_pool with mlu-ops (#2598)

* [Feature] Replace the implementation of deform_roi_pool with mlu-ops

* [Feature] Modify code

---------

Co-authored-by: budefei <budefei@cambricon.com>

* [Enhancement] ms_deform_attn performance optimization (#2616)

* ms_opt_v2

* ms_opt_v2_1

* optimize MultiScaleDeformableAttention ops for MLU

* ms_opt_v2_1

* [Feature] ms_deform_attn performance optimization V2

* [Feature] ms_deform_attn performance optimization V2

* [Feature] ms_deform_attn performance optimization V2

* [Feature] ms_deform_attn performance optimization V2

* [Feature] ms_deform_attn performance optimization V2

* [Feature] ms_deform_attn performance optimization V2

* [Feature] ms_deform_attn performance optimization V2

---------

Co-authored-by: dongchengwei <dongchengwei@cambricon.com>

* [Feature] Support NmsRotated with cambricon MLU backend (#2643)

* [Feature] Support NmsRotated with cambricon MLU backend

* [Feature] remove foolproofs in nms_rotated_mlu.cpp

* [Feature] fix lint in test_nms_rotated.py

* [Feature] fix kMLU not found in nms_rotated.cpp

* [Feature] modify mlu support in nms.py

* [Feature] modify nms_rotated support in ops.md

* [Feature] modify ops/nms.py

* [Enhance] Add a default value for MMCV_MLU_ARGS (#2688)

* add mlu_args

* add mlu_args

* Modify the code

---------

Co-authored-by: budefei <budefei@cambricon.com>

* [Enhance] Ignore mlu-ops files (#2691)

Co-authored-by: budefei <budefei@cambricon.com>

---------

Co-authored-by: ZShaopeng <108382403+ZShaopeng@users.noreply.github.com>
Co-authored-by: BinZheng <38182684+Wickyzheng@users.noreply.github.com>
Co-authored-by: liuduanhui <103939338+DanieeelLiu@users.noreply.github.com>
Co-authored-by: budefei <budefei@cambricon.com>
Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>
Co-authored-by: duzekun <108381389+duzekunKTH@users.noreply.github.com>
Co-authored-by: dongchengwei <dongchengwei@cambricon.com>
Co-authored-by: liuyuan1-v <125547457+liuyuan1-v@users.noreply.github.com>

ckirchhoff2021 committed:
Pick changes from 1.x branch (#2738)

zhouzaida committed:
Enhance issue templates (#2776)

zhouzaida committed:
[Fix] ops can not be built in github action (#2783)

* Fix CI

* Fix CI

* revert comments

* print cl

* use latest

* install wheel

* install wheel

* fix

* Update .github/workflows/pr_stage_test.yml

dflhw committed:
[Feature] Add the support of arf op for ascend device (#2792)

Ezra-Yu committed:
[Enhance] Raise a error if fails to read image (#2791)

* add assert

* Update mmcv/transforms/loading.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

* update comments

---------

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

zhouzaida committed:
[FIx] Build CUDA ops in CI (#2794)

HAOCHENYE committed:
[Enhancement] Make build_xxx_layer allow accepting a class type (#2782)

jiangjiang-coder66 committed:
Sync main with mmcv1.x branch (#2800)

MambaWong committed:
[Fix] Fix type hint in geometric (#2799)

kahakuka committed:
[Fix] Fix the support for ROCm (#2811)

qipengh committed:
[Refactor] Replace carafe op of MLU backend with mlu-ops (#2817)

DanieeelLiu committed:
[Refactor] Replace the implementation of psa_mask with mlu-ops. (#2810)

tudejiang79 committed:
[Feature] Add the support for rotated_feature_align with MLU (#2809)

tudejiang79 committed:
[Refactor] Repalce the implementation of roi_align_rotated with mlu-ops (#2808)

duzekunKTH committed:
[Refactor] Simplify the logic of sparse_conv (#2802)

dflhw committed:
[Fix] Keep the shape of iou op's parameter 2 smaller than parameter 1 (#2821)

dflhw committed:
[FIx] Fix arf op's write conflict when num_orientations is not 1 (#2824)

momo609 committed:
[Docs] Add npu build.md and update roi_pool_npu adpater (#2812)

dflhw committed:
Revert "[FIx] Fix arf op's write conflict when num_orientations is not 1 (#2824)" (#2825)

CokeDong committed:
[Feature] Support mmcv ext with DIOPI impl (#2790)

i-aki-y committed:
[Docs] Fix a typo in RandomResize docstring (#2829)

youkaichao committed:
[Feature] Add fast_conv_bn_eval option in ConvModule for fast validation and training in Eval mode (#2807)

zhouzaida committed:
Drop the outdated torch versions in CI (#2830)

wateryzephyr committed:
Correct mmengine version requirement (#2816)

Ginray committed:
[Fix] Fix deform_conv ops on Ascend NPU (#2832)

youkaichao committed:
[Enhance] Add ConvModule.turn_on_fast_conv_bn_eval to reduce repetitive code and dynamically bind conv during forward (#2835)

zhouzaida committed:
Bump version to 2.0.1 (#2831)

CokeDong committed:
[Enhancement] support dipu_mock_cuda=False in dipu for mmcv ext ops with cpu fallback (#2839)

* Support cuda_mock

* Fix

* Support roi_align

* Fix lint

* Support bbox

kahakuka committed:
Fix the support for ROCm (#2827)

zhouzaida committed:
Install wheel package in CI to support bdist_wheel command (#2851)

Danielmic committed:
[Feature] Add the implementation of dynamic_scatter with mlu-ops (#2847)

Danielmic committed:
[Feature] Add the implementation of diff_iou_rotated with mlu-ops (#2840)

long11111111 committed:
[Feature] Add support of points_in_polyogns for Ascend device (#2848)

momo609 committed:
[Fix] Fix roi_align npu bug (#2862)

MrShadowY committed:
[Feature] Add the support of BoxIouRotated op for ascend device (#2842)

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

Commit:
[Feature] Support modulated_deform_conv and deform_conv with cambricon MLU backend (#2823)

ClowDragon committed:
[Enhancement] Add torch mluops check before calling mluOpsxxx interface (#2871)

youkaichao committed:
rename fast_conv_bn_eval to efficient_conv_bn_eval (#2884)

youkaichao committed:
[Enhancement] Change the order of condition to make fx wok (#2883)

ClowDragon committed:
[Fix] Fix MLU code format (#2887)

CokeDong committed:
Adapt new dipu device (#2905)

qirun-uiuc committed:
[Refactor] Replace focal_loss_sigmoid op of MLU backend with mlu-ops (#2855)

qirun-uiuc committed:
[Refactor] Replace bbox_overlaps op of MLU backend with mlu-ops (#2879)

ClowDragon committed:
[Refactor] Replace roipoint_pool3d op of MLU backend with mlu-ops (#2875)

ClowDragon committed:
[Refactor] Replace tin_shift op of MLU backend with mlu-ops (#2910)

zstreet87 committed:
[Fix] Using PyTorch WARP_SHFL_DOWN macro for half support (#2843)

Commit:
Support pre released pytorch2.1.0 (#2865)

momo609 committed:
Support torch_npu 2.1 (#2909)

HarborYuan committed:
[Fix] Fix data race risk of `cache_randomness` (#2927)

ClowDragon committed:
[Fix] Update mlu-ops version (#2921)

momo609 committed:
[Fix] Fix setup.py for torch_npu v2.1 (#2928)

momo609 committed:
[Fix] Fix the support for nms_rotated in Ascend (#2931)

673958639 committed:
[Feature] Add the support for group_points ops with Ascend (#2935)

6Vvv committed:
Fix modulated_deform_conv for torch_npu v2.1 (#2941)

momo609 committed:
Use unified interface in Ascend kernel (#2945)

grimoire committed:
temporarily disable mps ops for torch2.1.0 (#2958)

zhouzaida committed:
Add torch2.1.0 checking in CI (#2955)

zhouzaida committed:
Bump version to 2.1.0 (#2959)

Ginray committed:
[Feature] Add the support of BallQuery op for Ascend device (#2963)

lihao7212148 committed:
[Feature] Add the support of three_interpolate op for Ascend device (#2962)

triple-mu committed:
Merge branch 'open-mmlab:main' into triplemu/fuse-dconv-bn

zhouzaida committed:
[Docs] Update version info (#2966)

triple-mu committed:
Merge branch 'open-mmlab:main' into triplemu/fuse-dconv-bn

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=2581) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=2581) before we can accept your contribution.<br/>**30** out of **31** committers have signed the CLA.<br/><br/>:white_check_mark: triple-mu<br/>:white_check_mark: jiangjiang-coder66<br/>:white_check_mark: ckirchhoff2021<br/>:white_check_mark: dflhw<br/>:white_check_mark: HAOCHENYE<br/>:white_check_mark: kahakuka<br/>:white_check_mark: DanieeelLiu<br/>:white_check_mark: duzekunKTH<br/>:white_check_mark: momo609<br/>:white_check_mark: youkaichao<br/>:white_check_mark: wateryzephyr<br/>:white_check_mark: i-aki-y<br/>:white_check_mark: Danielmic<br/>:white_check_mark: qirun-uiuc<br/>:white_check_mark: MrShadowY<br/>:white_check_mark: HarborYuan<br/>:white_check_mark: zstreet87<br/>:white_check_mark: Ginray<br/>:white_check_mark: 6Vvv<br/>:white_check_mark: grimoire<br/>:white_check_mark: lihao7212148<br/>:white_check_mark: long11111111<br/>:white_check_mark: 673958639<br/>:white_check_mark: zhouzaida<br/>:white_check_mark: CokeDong<br/>:white_check_mark: Ezra-Yu<br/>:white_check_mark: qipengh<br/>:white_check_mark: MambaWong<br/>:white_check_mark: tudejiang79<br/>:white_check_mark: ClowDragon<br/>:x: q.yao<br/><hr/>**q.yao** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmcv?pullRequest=2581) it.</sub>
